### PR TITLE
Add Write Side doc

### PR DESCRIPTION
**Pull request by EugeniyBurmistrov**

### CODE REVIEW

File: docs/basics/write-side.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
+---
+id: write-side
+title: Write Side
+---
+
+## Aggregates
+
+Commands are executed by objects that encapsulate domain logic. These objects are called Domain Objects.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md:8]: "encapsulate domain logic" Explain.

File: docs/basics/write-side.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+Usually Domain Objects are grouped into Aggregates. In a CQRS/ES app, an aggregate is a transaction boundary. This means that any given aggregate should be able to execute its commands without communicating with other aggregates.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: Why 'usually'?

File: docs/basics/write-side.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+Since the write side is used only to perform commands, your aggregate can be pretty slim, and only keep state that required for command execution.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: "can be pretty slim" Slim? Explain.

File: docs/basics/write-side.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+See Martin Fowler's definition for aggregates in the DDD paradigm: [https://martinfowler.com/bliki/DDD_Aggregate.html](https://martinfowler.com/bliki/DDD_Aggregate.html)
+
+In reSolve, an aggregate is a static object that contains a set of functions. of the following two kinds:
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: functions. of?

File: docs/basics/write-side.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+- [Command Handlers](#aggregate-command-handlers) - execute commands.
+
+Aggregate state is explicitly passed to all of these functions as an argument.
+
+## Aggregate ID
+
+Each aggregate should have a unique ID that is immutable during the aggregate's lifetime. An Aggregate ID should stay unique in the given event store, however we also recommend to keep it
+globally unique. It is also recommended that you generate aggregate IDs using [UUID v4](https://github.com/kelektiv/node-uuid#version-4) or [cuid](https://github.com/ericelliott/cuid) for distributed scalable apps.
+
+Please note that you have to generate a new Aggregate ID and send it with a command that creates a new aggregate.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: Please?

File: docs/basics/write-side.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+- [Projections](#aggregate-projection-function) - build aggregate state base from events.
+- [Command Handlers](#aggregate-command-handlers) - execute commands.
+
+Aggregate state is explicitly passed to all of these functions as an argument.
+
+## Aggregate ID
+
+Each aggregate should have a unique ID that is immutable during the aggregate's lifetime. An Aggregate ID should stay unique in the given event store, however we also recommend to keep it
+globally unique. It is also recommended that you generate aggregate IDs using [UUID v4](https://github.com/kelektiv/node-uuid#version-4) or [cuid](https://github.com/ericelliott/cuid) for distributed scalable apps.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: globally unique? Do you mean 'globally unique'?

Comment (dirk-pieterse) [reply docs/basics/write-side.md]: Do not use '-ing'. Rewrite.

File: docs/basics/write-side.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+## Sending a Command
+
+You can emit aggregate commands in the following cases:
+
+- [Sending commands from the client](#sending-commands-from-the-client)
+- [Emitting commands on the server](#emitting-commands-on-the-server)
+
+### Sending Commands From the Client
+
+The reSolve framework exposes an [HTTP API](api-reference.md#commands-http-api) that you can use to to send commands from the client side. Depending on the architecture of your web application's front-end, you can use this API directly or using the **Redux** binding mechanism provided by the **[resolve-redux](https://github.com/reimagined/resolve/tree/master/packages/core/resolve-redux)** library.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: Why 'use' and 'using'? Avoid -ing

File: docs/basics/write-side.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+You can send a command from the client side by sending a POST request to the following URL:
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: Why 'send' and 'sending'?

File: docs/basics/write-side.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+| Name              | Type   | Description                                                                                 |
+| ----------------- | ------ | ------------------------------------------------------------------------------------------- |
+| **aggregateId**   | string | The ID of an aggregate to which you are addressing the command                              |
+| **aggregateName** | string | The aggregate's name as defined in [config.app.js](../examples/shopping-list/config.app.js) |
+| **commandType**   | string | The command type supported by [the aggregate](../examples/shopping-list/common/aggregates)  |
+| **payload**       | object | The parameters that [the command](../examples/shopping-list/common/aggregates) accepts      |
+
+##### Example
+
+Use the following command to add an item to the **shopping-list** example.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: example. -> example:

File: docs/basics/write-side.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    "param2": value2,
+    ...
+    "paramN": valueN
+  }
+}
+```
+
+| Name              | Type   | Description                                                                                 |
+| ----------------- | ------ | ------------------------------------------------------------------------------------------- |
+| **aggregateId**   | string | The ID of an aggregate to which you are addressing the command                              |
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: This description does not read right. What are you trying to say?

File: docs/basics/write-side.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+| **aggregateName** | string | The aggregate's name as defined in [config.app.js](../examples/shopping-list/config.app.js) |
+| **commandType**   | string | The command type supported by [the aggregate](../examples/shopping-list/common/aggregates)  |
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: supported by? Explain.

File: docs/basics/write-side.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    "id":`date +%s`,
+    "text":"Learn reSolve API"
+  }
+}
+'
+```
+
+### Emitting Commands on the Server
+
+You can emit a command on the server side from a **[Saga](advanced-techniques.md#process-managers-sagas)** or **[API Handler](api-handlers.md)** using the **resolve.executeCommand** function as shown below:
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: Do not use 'using'

File: docs/basics/write-side.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+      payload: { title, text, link, userId, userName }
+    }
+  }
+  // ...
+}
+```
+
+## Aggregate Projection Function
+
+Projection functions are used to calculate an aggregate state based on the aggregate's events. A projection function receives a previous state and an event to be applied. It should return a new state based on the input.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: a previous state? any state? or a specific state?

File: docs/basics/write-side.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+
+Projection functions run for all events with the current aggregate ID. The resulting state is then passed to the corresponding [command handler](#aggregate-command-handlers).
+
+In addition to projection functions, a projection object should define an **Init** function. This function returns initial state of the aggregate.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: "returns initial state" No article?

File: docs/basics/write-side.md [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    voted: [],
+    comments: {}
+  })
+  // ...
+}
+```
+
+## Event Store
+
+All events returned by command handlers are saved to the event store. The reSolve framework writes events to the storage using one of the supported storage adapters.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: Do not use 'using'

File: docs/basics/write-side.md [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+}
+```
+
+Adapters for the following storage types are available out of the box:
+
+- [File or memory](https://github.com/reimagined/resolve/tree/master/packages/adapters/storage-adapters/resolve-storage-lite)
+- [MongoDB](https://github.com/reimagined/resolve/tree/master/packages/adapters/storage-adapters/resolve-storage-mongo)
+- [MySQL](https://github.com/reimagined/resolve/tree/master/packages/adapters/storage-adapters/resolve-storage-mysql)
+
+You can also write your own storage adapter and store events in any required way. To learn more about adapters, refer to the [Adapters](advanced-techniques.md#adapters) section of the reSolve documentation.
```
Comment (dirk-pieterse) [inline docs/basics/write-side.md]: in any required way?

Comment (dirk-pieterse) [reply docs/basics/write-side.md]: To learn more about adapters, refer to the [Adapters](advanced-techniques.md#adapters) section of the reSolve documentation -> Refer to the [Adapters](advanced-techniques.md#adapters) section of the reSolve documentation for more information about adapters.

File: docs/basics/write-side.md (hunk 14 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: Changed to "the initial state"

File: docs/basics/write-side.md (hunk 15 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: "The reSolve framework uses one of the supported storage adapters to write events to the storage."

File: docs/basics/write-side.md (hunk 12 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: "You can use the **resolve.executeCommand** function to emit a command on the server side from a **[Saga](advanced-techniques.md#process-managers-sagas)** or **[API Handler](api-handlers.md)**:"

File: docs/basics/write-side.md (hunk 9 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: done

File: docs/basics/write-side.md (hunk 13 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: changed to "the previous state"

File: docs/basics/write-side.md (hunk 16 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: > in any required way?

"You can also write your own storage adapter according to your requirements."

Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: > To learn more about adapters, refer to the [Adapters](advanced-techniques.md#adapters) section of the reSolve documentation -> Refer to the [Adapters](advanced-techniques.md#adapters) section of the reSolve documentation for more information about adapters.

Done

File: docs/basics/write-side.md (hunk 4 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: fixed

File: docs/basics/write-side.md (hunk 5 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: Removed "Please"

File: docs/basics/write-side.md (hunk 6 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: > globally unique? Do you mean 'globally unique'?

There seems to be no difference. Can you explain?

Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: > Do not use '-ing'. Rewrite.

It is also recommended that you use [UUID v4](https://github.com/kelektiv/node-uuid#version-4) or [cuid](https://github.com/ericelliott/cuid) to generate aggregate IDs for scalable apps.

File: docs/basics/write-side.md (hunk 10 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: "The ID of an aggregate to handle the command "

File: docs/basics/write-side.md (hunk 11 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: An aggregate supports (can handle) certain types of commands. Here, we are trying to say that you need to specify the command type, and that it should be one of the command types that the aggregate can handle.

File: docs/basics/write-side.md (hunk 1 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md:8]: Domain Logic is a term. From Wikipedia: "In computer software, business logic or domain logic is the part of the program that encodes the real-world business rules that determine how data can be created, stored, and changed."

Encapsulation is a fundamental principle of OOP: "A language mechanism for restricting direct access to some of the object's components."

When we say that Domain Objects encapsulate domain logic, we mean that they implement this logic and hide the implementation details from the outer world, like a black box in the engineering design sense.

File: docs/basics/write-side.md (hunk 2 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: Removed 'usually'

File: docs/basics/write-side.md (hunk 3 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: small/compact

File: docs/basics/write-side.md (hunk 7 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: Your application's frontend can use this API directly or through the **Redux** binding mechanism provided by the **[resolve-redux](https://github.com/reimagined/resolve/tree/master/packages/core/resolve-redux)** library.

File: docs/basics/write-side.md (hunk 8 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: You can send a command from the client side as a POST request to the following URL:

File: docs/basics/write-side.md (hunk 3 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: Compact

File: docs/basics/write-side.md (hunk 6 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: We recommend that you use UUID v4 or cuid to generate aggregate IDs for scalable apps.

File: docs/basics/write-side.md (hunk 7 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: Your application's frontend can use this API directly or through the Redux binding mechanism from the resolve-redux library.

File: docs/basics/write-side.md (hunk 8 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: Ok

File: docs/basics/write-side.md (hunk 10 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: The ID of an aggregate that should handle the command

File: docs/basics/write-side.md (hunk 11 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: The command type by [the aggregate] can handle?

File: docs/basics/write-side.md (hunk 12 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: Ok

File: docs/basics/write-side.md (hunk 16 above)
Comment (dirk-pieterse) [reply docs/basics/write-side.md]: You can also add your own storage adapter to store events.

File: docs/basics/write-side.md (hunk 3 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: done

File: docs/basics/write-side.md (hunk 6 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: done

File: docs/basics/write-side.md (hunk 7 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: done

File: docs/basics/write-side.md (hunk 10 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: done

File: docs/basics/write-side.md (hunk 11 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: Changed to your version

File: docs/basics/write-side.md (hunk 16 above)
Comment (EugeniyBurmistrov) [reply docs/basics/write-side.md]: done

**Pull request merged**